### PR TITLE
[Metric tensor tape] Adds a metric tensor function

### PR DESCRIPTION
**Pull request by josh146**
**Context:**

Follow up PR to #1012, adding in support for computing the metric tensor in tape mode.

**Description of the Change:**

* Adds the new `qml.metric_tensor` function, which transforms a QNode to produce a function that generates the metric tensor:
  Consider the following QNode:
  
  ```python
  dev = qml.device("default.qubit", wires=3)
  
  @qml.qnode(dev, interface="autograd")
  def circuit(weights):
      # layer 1
      qml.RX(weights[0, 0], wires=0)
      qml.RX(weights[0, 1], wires=1)
  
      qml.CNOT(wires=[0, 1])
      qml.CNOT(wires=[1, 2])
  
      # layer 2
      qml.RZ(weights[1, 0], wires=0)
      qml.RZ(weights[1, 1], wires=2)
  
      qml.CNOT(wires=[0, 1])
      qml.CNOT(wires=[1, 2])
      return qml.expval(qml.PauliZ(0) @ qml.PauliZ(1)), qml.expval(qml.PauliY(2))
  ```
  
  We can use the ``metric_tensor`` function to generate a new function, that returns the
  metric tensor of this QNode:
  
  ```pycon
  >>> met_fn = qml.metric_tensor(circuit)
  tensor([[0.25  , 0.    , 0.    , 0.    ],
          [0.    , 0.25  , 0.    , 0.    ],
          [0.    , 0.    , 0.0025, 0.0024],
          [0.    , 0.    , 0.0024, 0.0123]], requires_grad=True)
  ```
  
  The returned metric tensor is also fully differentiable, in all interfaces.
  For example, differentiating the ``(3, 2)`` element:
  
  ```pycon
  >>> grad_fn = qml.grad(lambda x: met_fn(x)[3, 2])
  >>> grad_fn(weights)
  array([[ 0.04867729, -0.00049502,  0.        ],
         [ 0.        ,  0.        ,  0.        ]])
  ```

 Differentiation is also supported using Torch, Jax, and TF.

* In addition the method `QNode.metric_tensor` is also provided. This is provided for backwards compatibility with the old UI, and should be eventually removed.

* The above function `qml.metric_tensor` internally calls the `qml.tape.transform.metric_tensor` function. This is a function of the following form:
  
  ```python
  def metric_tensor(tape):
      ...
      return metric_tensor_tapes, processing_fn
  ```

  That is, it computes the quantum tapes required to compute the metric tensor, as well as the final classical processing.

* To support the new metric tensor transform, `qml.math.block_diag` has been added.

* Finally, the test file [test_qnode_metric_tensor.py](https://github.com/PennyLaneAI/pennylane/blob/master/tests/qnodes/test_qnode_metric_tensor.py) has been duplicated to test for this new function.

**Benefits:**

* The metric tensor can now be computed in tape-mode

* The metric tensor is end-to-end differentiable on all interfaces, for all diff methods (including on hardware)

* Updated to provide a functional interface to the metric tensor

* The metric tensor computation is more efficient than the old non-tape mode version; the old numerical Gram-Schmidt approach has been removed.

**Possible Drawbacks:**

* This is a relatively efficient way of computing the block-diag metric tensor approximation on hardware with full autodiff support. However, there are [other techniques](https://arxiv.org/abs/2008.06517) for computing the _full_ metric tensor on hardware, that are a little more complicated. We should support this approach as well in the future.

* In the other direction, there are [efficient ways to compute the full metric tensor if we have access to a state-based simulator](https://arxiv.org/abs/2011.02991). We should also support this approach in addition in the future.

**Related GitHub Issues:** n/a

### CODE REVIEW

File: pennylane/math/tf_box.py [hunk 1]
```diff
@@ -68,6 +68,26 @@ def __len__(self):
...
+    @staticmethod
+    @wrap_output
+    def block_diag(values):
+        tensors = TensorFlowBox.unbox_list(values)
+        tensors = TensorFlowBox._coerce_types(tensors)
+        int_dtype = None
+
+        if tensors[0].dtype in (tf.int32, tf.int64):
+            int_dtype = tensors[0].dtype
+            tensors = [tf.cast(t, tf.float32) for t in tensors]
```
Comment (Thenerdstation) [inline pennylane/math/tf_box.py:80]: The casting feels a little awkward. I assume this is a TF limitation?

File: pennylane/tape/qnode.py [hunk 2]
```diff
@@ -445,6 +444,30 @@ def __call__(self, *args, **kwargs):
             return __import__(res_type_namespace).numpy.squeeze(res)
         return __import__(res_type_namespace).squeeze(res)
 
+    def metric_tensor(self, *args, diag_approx=False, only_construct=False, **kwargs):
+        """Evaluate the value of the metric tensor.
+
+        Args:
+            args (tuple[Any]): positional arguments
+            kwargs (dict[str, Any]): auxiliary arguments
+            diag_approx (bool): iff True, use the diagonal approximation
```
Comment (Thenerdstation) [inline pennylane/tape/qnode.py:453]:
```suggestion
            diag_approx (bool): Iff True, use the diagonal approximation

```

File: pennylane/tape/qnode.py [hunk 3]
```diff
@@ -725,6 +748,76 @@ def qfunc_decorator(func):
...
+
+    >>> grad_fn = qml.grad(lambda x: met_fn(x)[3, 2])
+    >>> grad_fn(weights)
+    array([[ 0.04867729, -0.00049502,  0.        ],
+           [ 0.        ,  0.        ,  0.        ]])
+    """
+
+    def _metric_tensor_fn(*args, **kwargs):
+        qnode.construct(args, kwargs)
+        metric_tensor_tapes, processing_fn = qml.tape.transforms.metric_tensor(
```
Comment (Thenerdstation) [inline pennylane/tape/qnode.py:811]: Love the functional approach here.

File: pennylane/tape/transforms/metric_tensor.py [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+                o.diagonalizing_gates()
+
+            qml.probs(wires=tape.wires)
+
+        metric_tensor_tapes.append(layer_tape)
+
+    def processing_fn(probs):
+        gs = []
+
+        for prob, obs, coeffs in zip(probs, obs_list, coeffs_list):
```
Comment (Thenerdstation) [inline pennylane/tape/transforms/metric_tensor.py:87]: I love that you use closures here to keep the `obs_list` and `coeffs_list` bookkeeping all within this single method. Really clean design!

File: pennylane/math/tf_box.py (hunk 1 above)
Comment (josh146) [reply pennylane/math/tf_box.py:80]: Yeah, it's super weird. I discovered while writing this that TF only allows creating block diagonal matrices from float/complex tensors, _not_ integer tensors 🙁

File: pennylane/tape/qnode.py [hunk 5]
```diff
@@ -725,6 +748,76 @@ def qfunc_decorator(func):
...
+            qnode.qtape, diag_approx=diag_approx
+        )
+
+        res = [t.execute(device=qnode.device) for t in metric_tensor_tapes]
```
Comment (josh146) [inline pennylane/tape/qnode.py:815]: @Thenerdstation I would love to be able to simply do
```python
res = dev.batch_execute(metric_tensor_tapes)
```
here. But this requires first moving the interfaces to the device!

File: pennylane/tape/transforms/metric_tensor.py (hunk 4 above)
Comment (josh146) [reply pennylane/tape/transforms/metric_tensor.py:87]: lol not going to lie was pleasantly surprised when I realized I could do this

File: pennylane/math/tf_box.py (hunk 1 above)
Comment (josh146) [reply pennylane/math/tf_box.py:80]: It's very unlikely the metric tensor will ever be integral, but at the same time, the `qml.math` should try and be as consistent across interfaces as possible to avoid future issues/bugs.

File: pennylane/math/fn.py [hunk 6]
```diff
@@ -163,6 +163,33 @@ def arcsin(tensor):
...
+    ...     np.array([[1, 2], [3, 4]]),
+    ...     torch.tensor([[1, 2, 3], [-1, -6, -3]]),
+    ...     torch.tensor(5)
+    ... ]
+    >>> qml.math.block_diag(t)
+    tensor([[ 1,  2,  0,  0,  0,  0],
+        [ 3,  4,  0,  0,  0,  0],
+        [ 0,  0,  1,  2,  3,  0],
+        [ 0,  0, -1, -6, -3,  0],
+        [ 0,  0,  0,  0,  0,  5]])
```
Comment (mariaschuld) [inline pennylane/math/fn.py:188]: Minor one, but will this format correctly?

File: pennylane/tape/qnode.py [hunk 7]
```diff
@@ -410,17 +420,6 @@ def construct(self, args, kwargs):
...
-        if self.interface == "autograd":
-            # HOTFIX: to maintain compatibility with core, here we treat
-            # all inputs that do not explicitly specify `requires_grad=False`
-            # as trainable. This should be removed at some point, forcing users
-            # to specify `requires_grad=True` for trainable parameters.
-            args = [
-                anp.array(a, requires_grad=True) if not hasattr(a, "requires_grad") else a
-                for a in args
-            ]
-
```
Comment (mariaschuld) [inline pennylane/tape/qnode.py:423]: I am wondering about this change? Isn't the tape constructed each time the qnode is called?

File: pennylane/tape/qnode.py (hunk 2 above)
Comment (mariaschuld) [reply pennylane/tape/qnode.py:453]: Wait, the convention is: capital if sentence with period, lower case if not, right?

File: pennylane/tape/qnode.py [hunk 8]
```diff
@@ -445,6 +444,30 @@ def __call__(self, *args, **kwargs):
...
+            only_construct (bool): Iff True, construct the circuits used for computing
+                the metric tensor but do not execute them, and return the tapes.
+
+        Returns:
+            array[float]: metric tensor
+        """
+        self.construct(args, kwargs)
+        metric_tensor_tapes, processing_fn = qml.tape.transforms.metric_tensor(
+            self.qtape, diag_approx=diag_approx
+        )
```
Comment (mariaschuld) [inline pennylane/tape/qnode.py:463]: Nice :) Like a diff rule

File: pennylane/tape/qnode.py (hunk 2 above)
Comment (josh146) [reply pennylane/tape/qnode.py:453]: yep!

File: pennylane/tape/transforms/metric_tensor.py [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+
+
+def _stopping_critera(obj):
+    if getattr(obj, "num_params", 0) == 0:
+        return True
+
+    if obj.name in ["RX", "RY", "RZ", "PhaseShift"]:
+        return True
+
+    return False
```
Comment (mariaschuld) [inline pennylane/tape/transforms/metric_tensor.py:28]: Mabye just use a `Rot` in the test to cover this line?

Comment (josh146) [reply pennylane/tape/transforms/metric_tensor.py:28]: Good idea!

File: pennylane/tape/transforms/metric_tensor.py [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+
+
+def metric_tensor(tape, diag_approx=False):
+    """Returns a list of tapes, and a classical processing function,
+    for computing the metric tensor of an input tape on hardware."""
+
+    # Only the RX, RY, RZ, and PhaseShift gates are supported.
```
Comment (mariaschuld) [inline pennylane/tape/transforms/metric_tensor.py:35]: But this is not a universal gate set, right? Isn't that a problem?

Comment (josh146) [reply pennylane/tape/transforms/metric_tensor.py:35]: Oops, I mean more out of the _parametrized_ gates available!

```suggestion
    # For parametrized operations, only the RX, RY, RZ, and PhaseShift gates are supported.
```

File: .github/CHANGELOG.md [hunk 11]
```diff
@@ -214,6 +214,56 @@
...
+          [0.    , 0.    , 0.0025, 0.0024],
+          [0.    , 0.    , 0.0024, 0.0123]], requires_grad=True)
+  ```
+
+  The returned metric tensor is also fully differentiable, in all interfaces.
+  For example, differentiating the `(3, 2)` element:
+
+  ```pycon
+  >>> grad_fn = qml.grad(lambda x: met_fn(x)[3, 2])
+  >>> grad_fn(weights)
```
Comment (mariaschuld) [inline .github/CHANGELOG.md:260]: To make this executable code, do you want to define the weights?

**Review by mariaschuld — APPROVED** [review]
Looks great @josh146.

What I reviewed: General logic. Played around with branch. 

What I did not review: Detailed idea of how metric tensor is constructed and how you get the test data.

File: .github/CHANGELOG.md (hunk 11 above)
Comment (josh146) [reply .github/CHANGELOG.md:260]: Yep, I just realised they were missing as well. Have just locally added them in!

File: pennylane/tape/qnode.py (hunk 8 above)
Comment (josh146) [reply pennylane/tape/qnode.py:463]: That was my inspiration 😆 

Funnily enough, in https://arxiv.org/abs/2008.06517, a 4-term parameter-shift-like rule is provided for computing every metric tensor element. So you get the full metric tensor on hardware, but with more evaluations (`4p` evaluations for `p` parameters) than the block diagaonal approximation (`L` evaluations, where `L` is the number of layers)

We should add this at some point (too much work to do it now).

File: pennylane/math/fn.py (hunk 6 above)
Comment (josh146) [reply pennylane/math/fn.py:188]: Fixed! Thanks for catching this one

File: pennylane/tape/transforms/metric_tensor.py (hunk 4 above)
Comment (Thenerdstation) [reply pennylane/tape/transforms/metric_tensor.py:93]: Closures are great. Feel free to use them more

File: pennylane/math/tf_box.py (hunk 1 above)
Comment (Thenerdstation) [reply pennylane/math/tf_box.py:81]: Easiest way to do that is to check if the type is integral and raise an error if so since we don't expect it to ever happen. If we ever hit it then we deal with it.

File: pennylane/tape/qnode.py [hunk 12]
```diff
@@ -779,6 +796,141 @@ def qfunc_decorator(func):
...
+                gate_params = classical_preprocessing(*args, **kwargs)
+
+            return tape.jacobian(gate_params, args)
+
+        return _jacobian
+
+
+def metric_tensor(_qnode, diag_approx=False, only_construct=False):
+    """metric_tensor(qnode, diag_approx=False)
+    Returns a function that evaluates the value of the metric tensor
```
Comment (Thenerdstation) [inline pennylane/tape/qnode.py:840]: "evaluates the value" sounds awkward to me. Mabye "Returns a function that returns the metric tensor of a given QNode"

File: tests/tape/transforms/test_metric_tensor.py [hunk 13]
```diff
@@ -0,0 +1,586 @@
...
+        loss = qml.metric_tensor(circuit)(weights_t)[2, 2]
+        loss.backward()
+
+        grad = weights_t.grad
+        expected = np.array([
+            np.cos(a) * np.cos(b) ** 2 * np.sin(a) / 2,
+            np.cos(a) ** 2 * np.sin(2 * b) / 4,
+            0
+        ])
+        assert np.allclose(grad, expected, atol=tol, rtol=0)
```
[the remark this answers is not in the document]
Comment (Thenerdstation) [reply tests/tape/transforms/test_metric_tensor.py:664]: +1

**Review by Thenerdstation — APPROVED** [review]
Looks good, only minor nitpicks.

File: pennylane/tape/qnode.py (hunk 12 above)
Comment (josh146) [reply pennylane/tape/qnode.py:840]:
```suggestion
    Returns a function that returns the value of the metric tensor
```
